_projects/coala_language_server.md

```diff
@@ -42,48 +42,86 @@ a rapidly growing list of editor and language communities:
 
 #### Features
 
 - Support linting for multiple programming languages.
-- Support 'textDocument/publishDiagnostics', 'textDocument/didChange' and 'textDocument/didSave' requests.
+- The current language server has implementations for:
+  - 'textDocument/didSave'
+  - 'textDocument/didChange'
+  - 'textDocument/publishDiagnostics'
+  and this project will improve and extend the support to at least
+  the following request types:
+  - 'textDocument/didOpen'
+  - 'textDocument/didClose'
+  - 'workspace/formatting'
+  - 'window/showMessageRequest'
 - (Advanced) Multiple plugins Support (e.g. Eclipse Che).
+- Some additional request types that are either out of the scope
+  of this project or are marginally redundant to build but which would
+  improve the user friendliness or functionality  of this language
+  server are:
+  - 'textDocument/willSaveWaitUntil'
+  - 'workspace/didChangeConfiguration'
+  - 'workspace/configuration'
+  - 'workspace/applyEdit'
+  - 'window/showMessage'
 
 #### Milestones
 
 ##### PREPARATION/BONDING
 
 - Design documentation about the language server exists.
   - Architecture of the language server
-  - Mechanism to support multiple languages.
+  - Mechanism to support multiple languages and requests.
 
 ##### CODING PHASE 1
 
 - A basic language server for coala exists.
   - Support linting for python.
-- Basic test suite exists.
-- Basic Documentation exists including asciinema showing the
+  - Supports primary request types:
+    - 'textDocument/didOpen'
+    - 'textDocument/didSave'
+    - 'textDocument/didClose'
+    - 'textDocument/publishDiagnostics'
+- Test suite exists with 100% coverage.
+- Basic Documentation exists including video showing the
   functionality working in any LS editor client.
+- Performance metrics of the standalone language server against
+  all the primary request types exist.
 
 ##### CODING PHASE 2
 
-- A basic language server for coala exists.
-  - Support multiple kinds of requests.
-- Basic test suite exists.
-- Basic Documentation exists including asciinema showing the
+- The Language server now supports:
+  - All mentioned request types.
+  - Multiple programming langauges (at least JS, Java, C, C++, C#).
+- Basic langserver client plugin exists.
+  - Supports all requests types supported by server.
+  - Supports all languages supported by server.
+- Basic test suite exists for the client plugin.
+- Basic Documentation exists including a video showing the
   functionality working in any LS editor client.
+- Performance metrics against all supported request types, languages
+  and the supported coala-ls client with langserver exists.
 
 ##### CODING PHASE 3
 
-- A language server for coala exists.
-  - Support multiple languages.
-- A full covering test suite exists.
-- Basic Documentation exists including asciinema showing the
+- An optimized language server for coala exists.
+  - Optimized entry point into coala.
+  - Optimized runtime config filtering.
```

Comment:
need more detail about what this is.

Probably a good idea to create a top level section above dedicated to Performance and Optimisation, as that will be a large part of this project.

Reply:
I think once I have the PR in and it gets reviewed I can expand on the details. However here is what I mean in general.

- Runtime coala configuration
  It is not necessary for certain bears to be run for each
  analysis cycle, especially during requests such as didChange.
  The language server can preprocess the configuration file to
  exclude any such bears.

- Language Server specific parameters in coafile
  The coafile configuration system can be updated to include
  some parameters specific to language server giving control
  to disable/enable sections to be run during lang server
  executions and also indicating the request type to enable
  the section for.

- Custom entrypoint into coala
  The overhead involved during executing coala can be reduced
  significantly by preprocessing configuration files and
  initialising bears which can then be reused for subsequent
  analysis cycles.